lib/puppet/util/windows/monkey_patches/dir.rb

```diff
@@ -0,0 +1,409 @@
+    to   = Puppet::Util::Windows::String.wide_string(string_check(to))
+    from = Puppet::Util::Windows::String.wide_string(string_check(from))
+
+    from_path = (0.chr * 1024).encode("UTF-16LE")
+
+    length = GetFullPathNameW(from, from_path.size, from_path, nil)
+
+    if length == 0 || length > from_path.size
+      raise SystemCallError.new("GetFullPathNameW", FFI.errno)
+    else
+      from_path.strip!
+    end
+
+    to_path = (0.chr * 1024).encode("UTF-16LE")
+
+    length = GetFullPathNameW(to, to_path.size, to_path, nil)
+
+    if length == 0 || length > to_path.size
+      raise SystemCallError.new("GetFullPathNameW", FFI.errno)
+    else
+      to_path.strip!
+    end
+
+    # You can create a junction to a directory that already exists, so
+    # long as it's empty.
+    unless CreateDirectoryW(to_path, nil)
+      if FFI.errno != ERROR_ALREADY_EXISTS
+        raise SystemCallError.new("CreateDirectoryW", FFI.errno)
+      end
+    end
+
+    begin
+      # Generic read & write + open existing + reparse point & backup semantics
+      handle = CreateFileW(
+        to_path,
+        GENERIC_READ | GENERIC_WRITE,
+        0,
+        nil,
+        OPEN_EXISTING,
+        FILE_FLAG_OPEN_REPARSE_POINT | FILE_FLAG_BACKUP_SEMANTICS,
+        0
+      )
+
+      if handle == INVALID_HANDLE_VALUE
+        raise SystemCallError.new("CreateFileW", FFI.errno)
+      end
+
+      target = "\\??\\".encode("UTF-16LE") + from_path
+
+      rdb = REPARSE_JDATA_BUFFER.new
+      rdb[:ReparseTag] = 2684354563 # IO_REPARSE_TAG_MOUNT_POINT
+      rdb[:ReparseDataLength] = target.bytesize + 12
+      rdb[:Reserved] = 0
+      rdb[:SubstituteNameOffset] = 0
+      rdb[:SubstituteNameLength] = target.bytesize
+      rdb[:PrintNameOffset] = target.bytesize + 2
+      rdb[:PrintNameLength] = 0
+      rdb[:PathBuffer] = target
+
+      bytes = FFI::MemoryPointer.new(:ulong)
+
+      begin
+        bool = DeviceIoControl(
+          handle,
+          CTL_CODE(FILE_DEVICE_FILE_SYSTEM, 41, METHOD_BUFFERED, 0),
+          rdb,
+          rdb[:ReparseDataLength] + rdb.header_size,
+          nil,
+          0,
+          bytes,
+          nil
+        )
+
+        error = FFI.errno
+
+        unless bool
+          RemoveDirectoryW(to_path)
+          raise SystemCallError.new("DeviceIoControl", error)
+        end
+      ensure
+        FFI::WIN32.CloseHandle(handle)
+      end
+    end
+
+    self
+  end
+
+  # Returns the path that a given junction points to. Raises an
+  # Errno::ENOENT error if the given path does not exist. Returns false
+  # if it is not a junction.
+  #
+  # Example:
+  #
+  #    Dir.mkdir('C:/from')
+  #    Dir.create_junction('C:/to', 'C:/from')
+  #    Dir.read_junction("c:/to") # => "c:/from"
+  #
+  def self.read_junction(junction)
+    return false unless Dir.junction?(junction)
+
+    junction = Puppet::Util::Windows::String.wide_string(string_check(junction))
+
+    junction_path = (0.chr * 1024).encode("UTF-16LE")
```

Comment:
Would be better to use `Encoding::UTF_16LE` constants. Same for other places where this is used.